lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,45 @@ class _RealmCore {
 
     if (syncError is ClientResetError) {
       syncConfig.clientResetHandler.onManualReset?.call(syncError);
-    } else {
-      syncConfig.syncErrorHandler(syncError);
+      return;
+    }
+
+    syncConfig.syncErrorHandler(syncError);
+  }
+
+  static void _callResetCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
   }
 
   static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
-    }
+    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset!;
+
     final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
```

Comment:
This is not related to this PR, but we need to be releasing these handles at the end of the callback. At the end, this should look something like:

```dart
final beforeHandle = RealmHandle._unowned(realmHandle);
try {
  final realm = RealmInternal.getUnowned(...);
  await beforeResetCallback(realm);
  unlock(true);
} catch (e) {
  // log e
  unlock(false);
} finally {
  beforeHandle.release();
}
```

Reply:
will fix

Reply:
Maybe we can achieve it now with something like:
    _callResetCallback(() async { 
        final beforeHandle = RealmHandle._unowned(realmHandle);
        final realm = RealmInternal.getUnowned(syncConfig, RealmHandle._unowned(realmHandle));
        await beforeResetCallback(realm), unlockCallbackFunc).whenComplete(()=>beforeHandle.release());
        }, unlockCallbackFunc);

Reply:
.whenComplete is not avaialble since this is `FutureOr`

Reply:
I will be fixing it

Reply:
fixed